File: ruby/lib/runapi/kling/resources/ai_avatar.rb
# frozen_string_literal: true

module RunApi
  module Kling
    module Resources
      # Kling AI avatar generation resource.
      # Generate talking avatar videos from images and audio.
      class AiAvatar
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/kling/ai_avatar"

        RESPONSE_CLASS = Types::AiAvatarResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedAiAvatarResponse

        def initialize(http)
          @http = http
        end

        # Generate an AI avatar video and wait until complete.
        #
        # @param params [Hash] AI avatar parameters
        # @return [RunApi::Kling::Types::CompletedAiAvatarResponse] completed task with videos
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Create an AI avatar generation task.
        #
        # @param params [Hash] AI avatar parameters
        # @return [RunApi::Kling::Types::AiAvatarResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get AI avatar task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Kling::Types::AiAvatarResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["avatar"], params)
        end
      end
    end
  end
end

File: python/src/runapi/kling/resources/ai_avatar.py
"""Kling AI avatar resource."""

from __future__ import annotations

from typing import Any, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import AiAvatarResponse, CompletedAiAvatarResponse


class AiAvatar(Resource):
    """Generate talking avatar videos from images and audio."""

    ENDPOINT = "/api/v1/kling/ai_avatar"

    RESPONSE_CLASS = AiAvatarResponse
    COMPLETED_RESPONSE_CLASS = CompletedAiAvatarResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Generate an AI avatar video and poll until it completes.

        Args:
            **params: AI avatar parameters (model, source_image_url, source_audio_url, prompt, ...).

        Returns:
            The completed task with videos.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create an AI avatar task and return immediately with an ``id``.

        Args:
            **params: AI avatar parameters (model, source_image_url, source_audio_url, prompt, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["avatar"], compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of an AI avatar task.

        Args:
            id: The task id.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)
